pytest/tests/test_contract_update.py

```diff
@@ -7,30 +7,119 @@
 Verifies that the update was executed.
 """
 
+import json
 import sys
 import time
 import pathlib
+import pytest
+from utils import MetricsTracker, load_binary_file
+
+from common_lib.constants import TGAS
 
 sys.path.append(str(pathlib.Path(__file__).resolve().parents[1]))
-from common_lib import shared, contracts
+from common_lib import shared
+from common_lib.contracts import V0_CONTRACT_PATH, CURRENT_CONTRACT_PATH, MIGRATE_CURRENT_CONTRACT_PATH, UpdateArgsV0, UpdateArgsV1, ConfigV1
 
 
-def test_propose_update():
-    # deploy v0
-    contract_v0 = contracts.load_mpc_contract_v0()
-    cluster = shared.start_cluster_with_mpc(2, 2, 1, contract_v0)
+@pytest.mark.parametrize("initial_contract_path,update_args", [
+    pytest.param(V0_CONTRACT_PATH,
+                 UpdateArgsV0(CURRENT_CONTRACT_PATH),
+                 id="update v0 to current"),
+    pytest.param(CURRENT_CONTRACT_PATH,
+                 UpdateArgsV1(code_path=MIGRATE_CURRENT_CONTRACT_PATH),
+                 id="update current code"),
+    pytest.param(CURRENT_CONTRACT_PATH,
+                 UpdateArgsV1(code_path=None,
+                              config=ConfigV1(max_num_requests_to_remove=2,
+                                              request_timeout_blocks=10)),
+                 id="update current config"),
+])
+def test_contract_update(initial_contract_path, update_args):
+    initial_contract = load_binary_file(initial_contract_path)
+    cluster = shared.start_cluster_with_mpc(2, 2, 1, initial_contract)
     # assert correct contract is deployed
-    cluster.assert_is_deployed(contract_v0)
+    cluster.assert_is_deployed(initial_contract)
     cluster.init_contract(threshold=2)
     # do some requests
-    cluster.send_and_await_signature_requests(2)
+    cluster.send_and_await_signature_requests(2, add_gas=150 * TGAS)
     # propose v1
-    contract_v1 = contracts.load_mpc_contract_v1()
-    cluster.propose_update(contract_v1)
+    cluster.propose_update(update_args.borsh_serialize())
     cluster.vote_update(0, 0)
     cluster.vote_update(1, 0)
     ## wait for the transaction to be included
     time.sleep(2)
     # assert v1 is now deployed
-    cluster.assert_is_deployed(contract_v1)
-    cluster.send_and_await_signature_requests(2)
+    if update_args.code is not None:
+        print("ensuring contract code is updated")
+        cluster.assert_is_deployed(update_args.code)
+    else:
+        print("ensuring config is updated")
+        expected_config = update_args.dump_json()
+        deployed_config = cluster.get_config()
+        assert deployed_config == expected_config
+    print("update completed")
+    cluster.send_and_await_signature_requests(2, add_gas=150 * TGAS)
+
+
+# In case a nonce conflict occurs during a vote_update call, rerun the test once.
```

Comment:
In my previous testing I was not able to produce nonce conflicts in localnet even under significant load (10s or 100s of txs). I was running the other pytest test_signature_request with high num_requests and didn't see any requests fail to be included. I believe this is because the local setup is too simple for transaction reordering to happen. So I wonder if there is some other underlying cause for flakiness observed here.

Reply:
In the timespan between the rpc request to fetch the current nonce value and until the update transaction is sent, the mpc participant already submitted other transactions and the nonce increased.

This improved significantly after using release version of near core, but I still get the error every now an then.

Still learning about keys and accounts. Does each access key have its own nonce and would it help to use a different access key for the update calls?